plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -0,0 +1,523 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { PluginDatabaseManager, UrlReader } from '@backstage/backend-common';
+import {
+  apiEntityV1alpha1Policy,
+  componentEntityV1alpha1Policy,
+  DefaultNamespaceEntityPolicy,
+  EntityPolicies,
+  EntityPolicy,
+  FieldFormatEntityPolicy,
+  groupEntityV1alpha1Policy,
+  locationEntityV1alpha1Policy,
+  makeValidator,
+  NoForeignRootFieldsEntityPolicy,
+  ReservedFieldsEntityPolicy,
+  SchemaValidEntityPolicy,
+  templateEntityV1alpha1Policy,
+  userEntityV1alpha1Policy,
+  Validators,
+} from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import lodash from 'lodash';
+import { Logger } from 'winston';
+import {
+  DatabaseEntitiesCatalog,
+  DatabaseLocationsCatalog,
+  EntitiesCatalog,
+  LocationsCatalog,
+} from '../catalog';
+import { DatabaseManager } from '../database';
+import {
+  AnnotateLocationEntityProcessor,
+  AzureApiReaderProcessor,
+  BitbucketApiReaderProcessor,
+  CodeOwnersProcessor,
+  EntityPolicyProcessor,
+  FileReaderProcessor,
+  GithubOrgReaderProcessor,
+  GithubReaderProcessor,
+  GitlabApiReaderProcessor,
+  GitlabReaderProcessor,
+  HigherOrderOperation,
+  HigherOrderOperations,
+  LocationProcessor,
+  LocationReaders,
+  LocationRefProcessor,
+  PlaceholderProcessor,
+  PlaceholderResolver,
+  StaticLocationProcessor,
+  UrlReaderProcessor,
+  YamlProcessor,
+} from '../ingestion';
+import { CatalogRulesEnforcer } from '../ingestion/CatalogRules';
+import { LdapOrgReaderProcessor } from '../ingestion/processors/LdapOrgReaderProcessor';
+import {
+  jsonPlaceholderResolver,
+  textPlaceholderResolver,
+  yamlPlaceholderResolver,
+} from '../ingestion/processors/PlaceholderProcessor';
+
+export type CatalogEnvironment = {
+  logger: Logger;
+  database: PluginDatabaseManager;
+  config: Config;
+  reader: UrlReader;
+};
+
+/**
+ * A builder that helps wire up all of the component parts of the catalog.
+ *
+ * The touch points where you can replace or extend behavior are as follows:
+ *
+ * - Reader processors can be added or replaced. These implement the
+ *   functionality of reading raw data from a location, in the form of an
+ *   entity definition file.
+ * - Parser processors can be added or replaced. These accept the raw data as
+ *   read by the previous processors and parse it into raw structured data
+ *   (for example, from a binary buffer containing yaml text, to a JS object
+ *   structure).
+ * - Placeholder resolvers can be replaced or added. These run on the raw
```

Comment:
It is an option for it. But it saves you from having to replace ALL processors just because of this

Reply:
Yep, just not liking where this would go as we add more built in processors, each with their own options. We can keep adding methods here, but it's at risk of becoming a bit of a huge mess. Tbh that might also be totally fine though, but I'm not feeling we have explored enough options here. Unless you have, and then please fill us in x)

Reply:
I mean explored other ways of doing the codification ^_>

Reply:
Nah.

Maybe we could have a couple fixed slots (that are run in order) that each have an array of processors that you can unshift/pop to? `pushProcessor(BEFORE_VALIDATE, new CodeownersProcessor())`

Open for suggestions

Reply:
Yeah was also thinking along those lines. Or perhaps more targeted replacement, like `replaceProcessor(CodeownersProcessor, new CodeownersProcessor(specialConfig))`, keyed by whatever we could imagine. We could add a base `ProcessorList extends Array` kinda class that perhaps hide some things and tailors it for use with processing, where it could even implement the processor interface itself

Reply:
Or processors themselves have a priority/order attribute that they get sorted by before running. So when you make a new processor you can say that its priority is BEFORE_VALIDATION or even otherProcessor.prio + 1 or something. For some well defined spaced apart values

Reply:
Yep, would love if we could come up with a model where we have ordered groups of processors, but the ordering within those groups is guaranteed to not matter at all.

Not too granular though. Main takeaway from writing CSS the past decade is that z-index sucks x)

Reply:
It's what leads to stuff like this https://material-ui.com/customization/z-index/ and we'd have to do something like it. Not exactly glorious